homeassistant/components/iperf3/__init__.py

```diff
@@ -0,0 +1,184 @@
+    CONF_HOST, CONF_PROTOCOL, CONF_HOSTS, CONF_SCAN_INTERVAL, ATTR_ENTITY_ID
+from homeassistant.helpers.discovery import async_load_platform
+from homeassistant.helpers.dispatcher import dispatcher_send
+from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
+from homeassistant.helpers.event import async_track_time_interval
+
+REQUIREMENTS = ['iperf3==0.1.10']
+
+DOMAIN = 'iperf3'
+DATA_UPDATED = '{}_data_updated'.format(DOMAIN)
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_DURATION = 'duration'
+CONF_PARALLEL = 'parallel'
+CONF_MANUAL = 'manual'
+
+DEFAULT_DURATION = 10
+DEFAULT_PORT = 5201
+DEFAULT_PARALLEL = 1
+DEFAULT_PROTOCOL = 'tcp'
+DEFAULT_INTERVAL = timedelta(minutes=60)
+
+ATTR_DOWNLOAD = 'download'
+ATTR_UPLOAD = 'upload'
+ATTR_VERSION = 'Version'
+
+UNIT_OF_MEASUREMENT = 'Mbit/s'
+
+SENSOR_TYPES = {
+    ATTR_DOWNLOAD: [ATTR_DOWNLOAD.capitalize(), UNIT_OF_MEASUREMENT],
+    ATTR_UPLOAD: [ATTR_UPLOAD.capitalize(), UNIT_OF_MEASUREMENT],
+}
+
+PROTOCOLS = ['tcp', 'udp']
+
+HOST_CONFIG_SCHEMA = vol.Schema({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+    vol.Optional(CONF_DURATION, default=DEFAULT_DURATION): vol.Range(5, 10),
+    vol.Optional(CONF_PARALLEL, default=DEFAULT_PARALLEL): vol.Range(1, 20),
+    vol.Optional(CONF_PROTOCOL, default=DEFAULT_PROTOCOL): vol.In(PROTOCOLS),
+})
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_HOSTS): vol.All(
+            cv.ensure_list, [HOST_CONFIG_SCHEMA]
+        ),
+        vol.Optional(CONF_MONITORED_CONDITIONS, default=list(SENSOR_TYPES)):
+            vol.All(cv.ensure_list, [vol.In(list(SENSOR_TYPES))]),
+        vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_INTERVAL): vol.All(
+            cv.time_period, cv.positive_timedelta
+        ),
+        vol.Optional(CONF_MANUAL, default=False): cv.boolean,
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+SERVICE_SCHEMA = vol.Schema({
+    vol.Optional(ATTR_ENTITY_ID, default=None): cv.string,
+})
+
+
+async def async_setup(hass, config):
+    """Set up the iperf3 component."""
+    import iperf3
+
+    hass.data[DOMAIN] = {}
+
+    conf = config[DOMAIN]
+    for host in conf[CONF_HOSTS]:
+        host_name = host[CONF_HOST]
+
+        client = iperf3.Client()
+        client.duration = host[CONF_DURATION]
+        client.server_hostname = host_name
+        client.port = host[CONF_PORT]
+        client.num_streams = host[CONF_PARALLEL]
+        client.protocol = host[CONF_PROTOCOL]
+        client.verbose = False
+
+        data = hass.data[DOMAIN][host_name] = Iperf3Data(hass, client)
+
+        if not conf[CONF_MANUAL]:
+            async_track_time_interval(
+                hass, data.update, conf[CONF_SCAN_INTERVAL]
+            )
+
+    def update(call):
+        """Service call to manually update the data."""
+        entity_id = call.data[ATTR_ENTITY_ID]
+        if entity_id in hass.data[DOMAIN]:
```

Comment:
Entity_id referencing won't work what I can tell. Probably only use host referencing.

Reply:
Yeah the old service call was using entity_id to represent host. This is a good time to change that.